src/Analyzers/CSharp/Analyzers/UseSystemThreadingLock/CSharpUseSystemThreadingLockDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,235 @@
+                {
+                    potentialLockFields.Remove(fieldReference);
+                    return;
+                }
+
+                // We did lock on this field, mark as such as its now something we'd def like to convert to a
+                // System.Threading.Lock if possible.
+                wasLockedSet.Add(fieldReference);
+                return;
+            }
+
+            // It's ok to assign to the field, as long as we're assigning a new lock object to it. e.g.  `_gate = new
+            // object()` is fine to continue converting over to System.Threading.Lock.  But an assignment of something
+            // else is not.
+            if (fieldReferenceOperation.Parent is IAssignmentOperation { Syntax: AssignmentExpressionSyntax assignmentSyntax } assignment &&
+                assignment.Target == fieldReferenceOperation &&
+                IsSystemObjectCreationExpression(assignmentSyntax.Right))
+            {
+                var operand = assignment.Value is IConversionOperation { Conversion: { Exists: true, IsImplicit: true } } conversion
+                    ? conversion.Operand
+                    : assignment.Value;
+
+                if (operand is IObjectCreationOperation { Arguments.Length: 0, Constructor.ContainingType.SpecialType: SpecialType.System_Object })
+                    return;
+            }
+
+            // Fine to use `nameof(someLock)` as that's not actually using the lock.
+            if (fieldReferenceOperation.Parent is INameOfOperation)
+                return;
+
+            // Note: More patterns can be added here as needed.
+
+            // This wasn't a supported case.  Immediately disallow conversion of this field.
+            potentialLockFields.Remove(fieldReference);
+        }, OperationKind.FieldReference);
+
+        context.RegisterSymbolEndAction(context =>
+        {
+            var cancellationToken = context.CancellationToken;
+
+            foreach (var lockField in potentialLockFields)
+            {
+                // Has to at least see this field locked on to offer to convert it to a Lock.
+                if (!wasLockedSet.Contains(lockField))
+                    continue;
+
+                // .Single is safe here as we confirmed there was only one DeclaringSyntaxReference in the field at the beginning of analysis.
+                var declarator = (VariableDeclaratorSyntax)lockField.DeclaringSyntaxReferences.Single().GetSyntax(cancellationToken);
+
+                context.ReportDiagnostic(DiagnosticHelper.Create(
+                    Descriptor,
+                    declarator.Identifier.GetLocation(),
+                    option.Notification,
+                    context.Options,
+                    additionalLocations: null,
+                    properties: null));
+            }
+        });
+
+        static bool IsSystemObjectCreationExpression(ExpressionSyntax expression)
```

Comment:
😕 Why operate on `ExpressionSyntax` when an operation context with full semantics is going to be available at all the locations of interest?

Reply:
it's not available at the declaration point of hte field unfortunately.  and you're not allowed to get a semantic model to analyze the field initializer :'(

Reply:
changed this up so we have the semantic model.